yetanotherco/aligned_layer
# feat (payment-service): lock and unlock user funds

Allow users to unlock and relock funds. Batcher checks if user funds are locked before allowing into batch.

# To test

Start everything normally

Go into batcher/aligned dir:
```cd batcher/aligned```

Then unlock user funds by running:
```
cast send 0x7969c5eD335650692Bc04293B07F5BF2e7A673C0 "unlock()" --rpc-url http://localhost:8545 --keystore ../../config-files/anvil.ecdsa.key.json
```

Then try to submit proofs to batcher (expect insufficient funds error) by running:
```
cargo run --release -- submit \
                --proving_system Groth16Bn254 \
                --proof ../../scripts/test_files/gnark_groth16_bn254_infinite_script/infinite_proofs/ineq_1_groth16.proof \
                --public_input ../../scripts/test_files/gnark_groth16_bn254_infinite_script/infinite_proofs/ineq_1_groth16.pub \
                --vk ../../scripts/test_files/gnark_groth16_bn254_infinite_script/infinite_proofs/ineq_1_groth16.vk \
                --repetitions 8 \
                --keystore_path ../../config-files/anvil.ecdsa.key.json
```

Then lock user funds by running:
```
cast send 0x7969c5eD335650692Bc04293B07F5BF2e7A673C0 "lock()" --rpc-url http://localhost:8545 --keystore ../../config-files/anvil.ecdsa.key.json
```

Then resubmit proofs and it should work correctly. 

The other thing to test is to try to withdraw funds before and after waiting 100 blocks (make sure to unlock funds beforehand):
```
cast send 0x7969c5eD335650692Bc04293B07F5BF2e7A673C0 "withdraw(uint256)" 100 --rpc-url http://localhost:8545 --keystore ../../config-files/anvil.ecdsa.key.json
```
